### PR TITLE
feat(lexers): add ASCII armored

### PR DESCRIPTION
**Pull request by scop**
Test data to follow later once questions raised have been sorted out.

### CODE REVIEW

File: pygments/lexers/asc.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+__all__ = ['AscLexer']
+
+
+class AscLexer(RegexLexer):
+    """
+    Lexer for ASCII armored files, containing `-----BEGIN/END ...-----` wrapped base64 data.
+    """
+    name = 'ASCII armored'
+    aliases = ['asc', 'pem']
```
Comment (scop) [inline pygments/lexers/asc.py:23]: Should we add some more here?

File: pygments/lexers/asc.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    filenames = [
+        '*.asc',  # PGP; *.gpg, *.pgp, and *.sig too, but those can be binary
+        '*.pem',  # X.509; *.cer, *.crt, *.csr, and key etc too, but those can be binary
+        'id_dsa', 'id_ecdsa', 'id_ecdsa_sk', 'id_ed25519', 'id_ed25519_sk', 'id_rsa',  # SSH private keys
+    ]
+    mimetypes = ['application/pgp-keys', 'application/pgp-encrypted', 'application/pgp-signature'],
```
Comment (scop) [inline pygments/lexers/asc.py:29]: There are some MIME types around for X.509 files, but they don't differentiate ASCII armored and binary encodings.

File: pygments/lexers/asc.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    tokens = {
+        'root': [
+            (r'\s+', Text),
+            (r'^-----BEGIN [^\n]+-----$', Generic.Heading, 'data'),
+            (r'\S+', Comment),
+        ],
+        'data': [
+            (r'\s+', Text),
+            (r'^([^:]+)(:)([ \t]+)(.*)', bygroups(Name.Attribute, Operator, Text, String)),
```
Comment (scop) [inline pygments/lexers/asc.py:41]: I wonder if we should use a different token type for the value here and the base64 data (`\S+` below), both are `String` now.

File: pygments/lexers/asc.py (hunk 1 above)
Comment (Anteru) [reply pygments/lexers/asc.py:23]: As long as they're not ambiguous it's fine.

File: pygments/lexers/asc.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+from pygments.lexer import RegexLexer, bygroups
+from pygments.token import Comment, Generic, Name, Operator, Text, String
+
+__all__ = ['AscLexer']
+
+
+class AscLexer(RegexLexer):
+    """
+    Lexer for ASCII armored files, containing `-----BEGIN/END ...-----` wrapped base64 data.
```
Comment (Anteru) [inline pygments/lexers/asc.py:20]: Please add `.. versionadded:: 2.10`

**Comment by Anteru** [comment]
Thanks for starting to work on this! Please note that any new lexer should come with an example file and golden test output -- can you please add that?

**Comment by scop** [comment]
Re test data: sure, see initial comment.

File: pygments/lexers/asc.py (hunk 1 above)
Comment (scop) [reply pygments/lexers/asc.py:23]: I was thinking about certificates and the like, but as/if we're not adding filenames and mimetypes for them, I think it's ok to leave them out from here, too.

File: pygments/lexers/asc.py (hunk 4 above)
Comment (scop) [reply pygments/lexers/asc.py:20]: Done.

File: pygments/lexers/asc.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        '*.pem',  # X.509; *.cer, *.crt, *.csr, and key etc too, but those can be binary
+        'id_dsa', 'id_ecdsa', 'id_ecdsa_sk', 'id_ed25519', 'id_ed25519_sk', 'id_rsa',  # SSH private keys
+    ]
+    mimetypes = ['application/pgp-keys', 'application/pgp-encrypted', 'application/pgp-signature']
+
+    flags = re.MULTILINE
+
+    tokens = {
+        'root': [
+            (r'\s+', Text),
```
Comment (Anteru) [inline pygments/lexers/asc.py:37]: That should be whitespace, right?

File: pygments/lexers/asc.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            (r'^-----BEGIN [^\n]+-----$', Generic.Heading, 'data'),
+            (r'\S+', Comment),
+        ],
+        'data': [
+            (r'\s+', Text),
```
Comment (Anteru) [inline pygments/lexers/asc.py:42]: Same here, whitespace.

File: pygments/lexers/asc.py (hunk 3 above)
Comment (Anteru) [reply pygments/lexers/asc.py:43]: No strong preference.

**Review by Anteru — CHANGES_REQUESTED** [review]

File: pygments/lexers/asc.py (hunk 5 above)
Comment (scop) [reply pygments/lexers/asc.py:37]: I don't know, seems a lot of lexers mark whitespace as Text, and significantly fewer as Whitespace.

```shellsession
$ git grep -E "'\\\s[+*]', Text\)" | wc -l
272
$ git grep -E "'\\\s[+*]', .*Whitespace" | wc -l
33
```

But I'm fine either way, just let me know.

Comment (Anteru) [reply pygments/lexers/asc.py:37]: Yeah, that's because we haven't paid enough attention to this previously :)

Comment (scop) [reply pygments/lexers/asc.py]: Ok, changed to Whitespace.

**Review by Anteru — APPROVED** [review]

**Pull request merged**